File: 04-linked-list/singly-linked-list/singlylinkedlist.js
class LinkedListNode {
  constructor(data) {
    this.data = data; // the data value stored in the node
    this.next = null; // a reference to the next node in the list
  }
}
class LinkedListIterator {
  constructor(node) {
    this.currentNode = node;
  }

  data() {
    if (this.currentNode == null) return null;
    return this.currentNode.data;
  }

  next() {
    if (this.currentNode != null) {
      this.currentNode = this.currentNode.next;
    }
    return this;
  }

  current() {
    return this.currentNode;
  }
}
class SingleLinkedList {
  constructor() {
    this.length = 0;
    this.head = null;
    this.tail = null;
  }

  addHead(node) {
    this.head = node;
    this.tail = node;
  }

  // canInsert(_data) {
  //   if (this.unique && this.isExist(_data)) {
  //     console.log(_data, "already exist!");
  //     return false;
  //   } else {
  //     return true;
  //   }
  // }

  insertLast(_data) {
    // if (!canInsert(_data)) return;

    const newNode = new LinkedListNode(_data);
    if (this.head === null) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      this.tail.next = newNode;
      this.tail = newNode;
    }
    this.length++;
  }

  insertAfter(node, _data) {
    //if (!canInsert(_data)) return;

    const newNode = new LinkedListNode(_data);
    newNode.next = node.next;
    node.next = newNode;
    if (newNode.next === null) {
      this.tail = newNode;
    }
    this.length++;
  }

  deleteNode(node) {
    if (this.head === this.tail) {
      this.head = null;
      this.tail = null;
    } else if (this.head === node) {
      this.head = node.next;
    } else {
      const parentNode = this.findParent(node);

      if (this.tail === node) {
        this.tail = parentNode;
      } else {
        parentNode.next = node.next;
      }
    }
    this.length--;
  }

  deleteNodeByData(_data) {
    const node = this.find(_data);
    if (node === null) {
      return;
    }
    this.deleteNode(node);
  }

  deleteHead() {
    if (this.head === null) {
      return;
    }
    this.deleteNode(this.head);
  }

  insertBefore(node, _data) {
    //if (!canInsert(_data)) return;
    const newnode = new LinkedListNode(_data);
    newnode.next = node;

    const parentNode = this.findParent(node);

    if (parentNode === null) {
      this.head = newnode;
    } else {
      parentNode.next = newnode;
    }
    this.length++;
  }

  find(_data) {
    for (let itr = this.begin(); itr.current() !== null; itr.next()) {
      if (itr.data() === _data) {
        return itr.current();
      }
    }
    return null;
  }

  findParent(node) {
    for (let itr = this.begin(); itr.current() !== null; itr.next()) {
      if (itr.current().next === node) {
        return itr.current();
      }
    }
    return null;
  }

  getLengthItr() {
    let count = 0;
    for (let itr = this.begin(); itr.current() !== null; itr.next()) {
      count++;
    }
    return count;
  }

  size() {
    return this.length;
  }

  printList() {
    let output = "";
    for (let itr = this.begin(); itr.current() !== null; itr.next()) {
      output += itr.data() + " -> ";
    }
    console.log(output);
  }

  sum() {
    let sum = 0;
    for (let itr = this.begin(); itr.current() !== null; itr.next()) {
      sum += itr.data();
    }
    return sum;
  }
  begin() {
    const itr = new LinkedListIterator(this.head);
    return itr;
  }
}

module.exports = {
  SingleLinkedList: SingleLinkedList,
  LinkedListNode: LinkedListNode,
  LinkedListIterator: LinkedListIterator,
};

File: 04-linked-list/singly-linked-list/singly_linked_list.py
# Define the singly linked list node class
class LinkedListNode:

  def __init__(self, data):
    self.data = data
    self.next = None


# Define the linked list iterator class
class LinkedListIterator:

  def __init__(self, node=None):
    self.currentNode = node

  # Return the current node's data
  def data(self):
    return self.currentNode.data

  # Advance to the next node and return the iterator
  def next(self):
    self.currentNode = self.currentNode.next
    return self

  # Return the current node
  def current(self):
    return self.currentNode


# Define the single linked list class
class SingleLinkedList:

  def __init__(self):
    self.length = 0
    self.head = None
    self.tail = None

  # Add a node to the head of the list
  def addHead(self, node):
    self.head = node
    self.tail = node

  # Insert a new node at the end of the list
  def insertLast(self, data):
    newNode = LinkedListNode(data)
    if self.head is None:
      self.head = newNode
      self.tail = newNode
    else:
      self.tail.next = newNode
      self.tail = newNode
    self.length += 1

  # Insert a new node after a given node
  def insertAfter(self, node, data):
    newnode = LinkedListNode(data)
    newnode.next = node.next
    node.next = newnode
    if newnode.next is None:
      self.tail = newnode
    self.length += 1

  # Delete a given node from the list
  def deleteNode(self, node):
    if self.head == self.tail:
      self.head = None
      self.tail = None
    elif self.head == node:
      self.head = node.next
    else:
      parentNode = self.findParent(node)

      if self.tail == node:
        self.tail = parentNode
      else:
        parentNode.next = node.next
    self.length -= 1

  # Delete the first node in the list
  def deleteHead(self):
    if self.head is None:
      return
    self.deleteNode(self.head)

  # Insert a new node before a given node
  def insertBefore(self, node, data):
    newnode = LinkedListNode(data)
    newnode.next = node

    parentNode = self.findParent(node)

    if parentNode is None:
      self.head = newnode
    else:
      parentNode.next = newnode
    self.length += 1

  # Find a node with the given data value
  def find(self, data):
    itr = self.begin()
    while itr.current() is not None:
      if itr.data() == data:
        return itr.current()
      itr.next()
    return None

  # Find the parent of a given node
  def findParent(self, node):
    itr = self.begin()
    while itr.current() is not None:
      if itr.current().next == node:
        return itr.current()
      itr.next()
    return None

  # Return the length of the list using an iterator
  def getLengthItr(self):
    count = 0
    itr = self.begin()
    while itr.current() is not None:
      count += 1
      itr.next()
    return count

  # Return the length of the list
  def Length(self):
    return self.length

  # Print the contents of the list
  def printList(self):
    itr = self.begin()
    while itr.current() is not None:
      print(itr.data(), "->", end=" ")
      itr.next()

  def printList(self):
    itr = self.begin()
    while itr.current() != None:
      print(itr.data(), end=" -> ")
      itr.next()
    print()

  def sum(self):
    sum = 0
    itr = self.begin()
    while itr.current() != None:
      sum += itr.data()
      itr.next()
    return sum

  def begin(self):
    itr = LinkedListIterator(self.head)
    return itr
